### PR TITLE
Add a workaround for SQL Client bug involving parameters of type date and time

### LINKED ISSUE #8894: SQL date type being cast to datetime
When using the SQL 'date' type I should be able to insert a date prior to 1753, but I receive the following exception.

```
Microsoft.EntityFrameworkCore.DbUpdateException occurred
  HResult=0x80131500
  Message=An error occurred while updating the entries. See the inner exception for details.
  Source=<Cannot evaluate the exception source>
  StackTrace:
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(Tuple`2 parameters)
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](Func`2 operation, Func`2 verifySucceeded, TState state)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(IReadOnlyList`1 entriesToSave)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at ConsoleApp1.Program.Main(String[] args) in C:\ConsoleApp1\Program.cs:line 14

Inner Exception 1:
SqlTypeException: SqlDateTime overflow. Must be between 1/1/1753 12:00:00 AM and 12/31/9999 11:59:59 PM.
```

### Steps to reproduce

```c#
using Microsoft.EntityFrameworkCore;
using System;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            using (var db = new ApplicationContext())
            {
                var item = new Item { Date = new DateTime(1605, 1, 15) };
                db.Add(item);
                db.SaveChanges();
            }
        }
    }

    public class ApplicationContext : DbContext
    {
        public DbSet<Item> Items { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=TempDateTest;Trusted_Connection=True;");
        }

        protected override void OnModelCreating(ModelBuilder builder)
        {
            base.OnModelCreating(builder);

            builder.Entity<Item>()
                .Property(m => m.Date)
                .ForSqlServerHasColumnType("date");
        }
    }

    public class Item
    {
        public int Id { get; set; }
        public DateTime Date { get; set; }
    }
}
```

### Further technical details
EF Core version: 1.1.2
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 8.1
IDE: Visual Studio 2017

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #8894

### CODE REVIEW

File: test/EFCore.SqlServer.FunctionalTests/BuiltInDataTypesSqlServerTest.cs [hunk 1]
```diff
@@ -330,7 +330,7 @@ public virtual void Can_insert_and_read_back_all_mapped_data_types()
 @p3='True'
 @p4='Your' (Nullable = false) (Size = 8000) (DbType = AnsiString)
 @p5='strong' (Nullable = false) (Size = 8000) (DbType = AnsiString)
-@p6='01/02/2015 10:11:12' (DbType = DateTime)
+@p6='01/02/2015 10:11:12' (DbType = Date)
```
Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/BuiltInDataTypesSqlServerTest.cs]: Why no SQL is using year 1605?

File: src/EFCore.SqlServer/Storage/Internal/SqlServerTimeSpanTypeMapping.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        protected override void ConfigureParameter(DbParameter parameter)
+        {
+            base.ConfigureParameter(parameter);
+
+            // Workaround for a SQLClient bug
+            if (DbType == System.Data.DbType.Time)
+            {
+                ((SqlParameter)parameter).SqlDbType = SqlDbType.Time;
```
Comment (smitpatel) [inline src/EFCore.SqlServer/Storage/Internal/SqlServerTimeSpanTypeMapping.cs]: Is this access null ref safe?

File: test/EFCore.SqlServer.FunctionalTests/BuiltInDataTypesSqlServerTest.cs (hunk 1 above)
Comment (AndriySvyryd) [inline test/EFCore.SqlServer.FunctionalTests/BuiltInDataTypesSqlServerTest.cs]: Because the test I changed doesn't validate SQL

Comment (smitpatel) [inline test/EFCore.SqlServer.FunctionalTests/BuiltInDataTypesSqlServerTest.cs]: Add SQL validation to the test if it is static.

Comment (AndriySvyryd) [inline test/EFCore.SqlServer.FunctionalTests/BuiltInDataTypesSqlServerTest.cs]: The test already asserts round-tripping, SQL validation is redundant

File: src/EFCore.SqlServer/Storage/Internal/SqlServerTimeSpanTypeMapping.cs (hunk 2 above)
Comment (AndriySvyryd) [inline src/EFCore.SqlServer/Storage/Internal/SqlServerTimeSpanTypeMapping.cs]: You mean `parameter`? Yes, it's never null

**Pull request merged**